### DISCUSSION TITLE
Create relationships between Dim and Fact

### DISCUSSION THREAD

**Discussion by thodegaard** [TabularEditor/TabularEditor3#532] [🙋♀️TE3 Q&A] [answered]
Hi,

I am trying to create a macro in TE3 that automatically detects and creates relationships between my facts and dimensions tables, based on column names that starts with "DW_SK_", however it does not do anything.
The context is set to tables and my macro looks like this:

```csharp
var keyPrefix = "DW_SK_";

// Loop through all currently selected tables (assumed to be fact tables):
foreach(var fact in Selected.Tables)
{
    // Loop through all SK columns on the current table:
    foreach(var factColumn in fact.Columns.Where(c => c.Name.StartsWith(keyPrefix)))
    {
        // Find the dimension table corresponding to the current SK column:
        var dim = Model.Tables.FirstOrDefault(t => factColumn.Name.StartsWith(keyPrefix + t.Name));
        if(dim != null)
        {
            // Find the key column on the dimension table:
            var dimColumn = dim.Columns.FirstOrDefault(c => factColumn.Name.EndsWith(c.Name));
            if(dimColumn != null)
            {
                // Check whether a relationship already exists between the two columns:
                if(!Model.Relationships.Any(r => r.FromColumn == factColumn && r.ToColumn == dimColumn))
                {
                    // If relationships already exists between the two tables, new relationships will be created as inactive:
                    var makeInactive = Model.Relationships.Any(r => r.FromTable == fact && r.ToTable == dim);

                    // Add the new relationship:
                    var rel = Model.AddRelationship();
                    rel.FromColumn = factColumn;
                    rel.ToColumn = dimColumn;
                    factColumn.IsHidden = true;
                    if(makeInactive) rel.IsActive = false;
                }
            }
        }
    }
}
```

Does anyone know, what I do wrong?

Br.,
Thomas

**Comment by otykier** [comment]
Hi Thomas
Did you select one or more fact tables in the TOM Explorer, before executing the script/running the macro?

Also, assuming you have a column named `DW_SK_Product` in your fact table. Is the corresponding dimension table called `Product`? (Note that casing must match exactly)

**Reply by thodegaard** [reply]
Hi Daniel,

Yes, I am selecting the relevant fact table(s) before running the macro, but the corresponding dimension table is called DimProduct, so that might be, why it does not work. Do you know what to do with the script? I guess, I should modify 
"var dim = Model.Tables.FirstOrDefault(t => factColumn.Name.StartsWith(keyPrefix + t.Name));
        if(dim != null)"

**Reply by otykier** [reply, marked as answer]
Do you intend to keep the dimension table named that way? Otherwise, I recommend fixing the dimension table names before running the script (you can use batch rename to get rid of all the "Dim" prefixes in one go).

If you want to keep the Dim prefix, change this line:

```csharp
// Find the dimension table corresponding to the current SK column:
var dim = Model.Tables.FirstOrDefault(t => t.Name.StartsWith("Dim") && factColumn.Name.StartsWith(keyPrefix + t.Name.Substring(3)));
```

**Reply by thodegaard** [reply]
You are right. It is better to change the table names and get rid of Dim. Ohh, really cool feature with the batch rename! Did know that :D
Thanks, a lot Daniel.
